dev/src/bundle.ts

```diff
@@ -84,16 +84,33 @@ export class BundleBuilder {
     return this;
   }
 
-  private addBundledDocument(snap: DocumentSnapshot): void {
-    const docProto = snap.toDocumentProto();
-    this.documents.set(snap.id, {
-      document: snap.exists ? docProto : undefined,
-      metadata: {
-        name: docProto.name,
-        readTime: snap.readTime.toProto().timestampValue,
-        exists: snap.exists,
-      },
-    });
+  private addBundledDocument(snap: DocumentSnapshot, queryName?: string): void {
+    const originalDocument = this.documents.get(snap.id);
+    const originalQueries = originalDocument?.metadata.queries;
+
+    // Update with document built from `snap` because it is newer.
+    if (
+      !originalDocument ||
+      Timestamp.fromProto(originalDocument.metadata.readTime!) < snap.readTime
+    ) {
+      const docProto = snap.toDocumentProto();
+      this.documents.set(snap.id, {
+        document: snap.exists ? docProto : undefined,
+        metadata: {
+          name: docProto.name,
+          readTime: snap.readTime.toProto().timestampValue,
+          exists: snap.exists,
+        },
+      });
+    }
+
+    // Update `queries` to include both original and `queryName`.
+    const newDocument = this.documents.get(snap.id)!;
+    newDocument.metadata.queries = originalQueries || [];
+    if (queryName) {
+      newDocument.metadata.queries!.push(queryName);
+    }
+
     if (snap.readTime > this.latestReadTime) {
       this.latestReadTime = snap.readTime;
     }
```

Comment:
Can you change the error message right below to:

'Query name conflict: "${name}" has already been added.'

Reply:
Done.